fix(chain): improve chain scheduler robustness and eth tx monitoring

rvagg opened the pull request:
Still working on https://github.com/filecoin-project/curio/issues/548, it's still not fixed. And my latest diagnosis from looking at the logs is that the `MessageWatcherEth` just stops being called at some point so it doesn't even have a chance to run. Which leaves us looking at `CurioChainSched` as a potential culprit. But I don't have any useful logs that tell me what might be wrong.

So this PR:

* Adds a mutex around `callbacks`, which is not a likely culprit since I think all listeners are added at construction time, synchronously, but this seems like a good thing to do because there's no docs or guards that would stop new callbacks being added async
* Lots more logging so we get insight into its operation
* A secondary watcher ticker that will fire every 60 seconds to make sure we're still getting notifications, and if we haven't had a notification in 60 seconds then assume something is wrong and set up a new channel

magik6k commented on lib/chainsched/chain_sched.go:
```diff
@@ -41,7 +48,8 @@ func (s *CurioChainSched) AddHandler(ch UpdateFunc) error {
 		return xerrors.Errorf("cannot add handler after start")
 	}
-
+	s.callbacksLk.Lock()
+	defer s.callbacksLk.Unlock()
```
I'd put the log above started. Even though I'm 99% sure that all hook adds happen in construction, if we want to use the lock as a guard, for it to actually do anything it has to be above the `if started`

magik6k reviewed (commented):
Generally good to be defensive here so makes sense to me; One comment to address then it should be good.

Would be good to get in the habit of writing tests for the more critical components like this, but not going to block on that..

Fwiw in the years of having curio in production I've never had seen a single issue originating from the chain watcher so your problem is a little suspect. I'd suggest looking at `http://node-addr:12300/debug/pprof/goroutine?debug=1`, I'd guess that the watcher is running, just stuck somewhere (e.g. long-running postgres operation because e.g. some query is missing some index, or some index is a hash index (yb default and a big footgun) and not sorted - which has historically been the case quite often)

rvagg commented:
> that the watcher is running, just stuck somewhere

so after much tinkering and eventually deploying https://github.com/filecoin-project/curio/pull/553 this has fixed the problem. The call to ethclient's `TransactionByHash` just locks up and doesn't return. On the lotus side we have an equally weird error and my hunch is that lotus is returning an error in a form that ethclient isn't built to recognise so we don't get a return. On the lotus side I suspect this might be to do with a low-resource machine hitting splitstore compaction and then having some trouble properly processing the request resulting in an error. So, probably things to fix on the lotus side, but also here we need to be more resillient.

I'll combine the two into one, PR, remove some of the cruft and get it in better shape for merging.

rvagg commented:
@magik6k I incorporated the changes in #553, added a bunch of tests. The line that actually fixes the problem is https://github.com/filecoin-project/curio/blob/c52f983e5cc816074c6db75f07d44491285f9a6b/tasks/message/watch_eth.go#L130.

Just timing out the ethclient calls is enough to keep it all running. But I've left the other pieces in place because IMO they add a bit more robustness. I've had to do some restructuring to make some of these testable, unfortunately it increases the noise of the diff but they should be an improvement overall in being able to add tests in the future.

#553 is identical to this now but it's for the `synapse` branch, while this is for `main`.

magik6k commented:
Looks good now, needs conflicts resolved

rvagg commented:
welp, most of this came in with the synapse branch merge, but there's some review items in here now to land

rjan90 commented:
@magik6k can you sanity check the latest commit here after you approved? And maybe push the ⛙ button if all is okay?
